Lib/concurrent/futures/thread.py

```diff
@@ -215,10 +215,25 @@ def _initializer_failed(self):
                 if work_item is not None:
                     work_item.future.set_exception(BrokenThreadPool(self._broken))
 
-    def shutdown(self, wait=True):
+    def shutdown(self, wait=True, cancel_futures=False):
         with self._shutdown_lock:
             self._shutdown = True
             self._work_queue.put(None)
+            if cancel_futures:
+                # Drain all work items from the queue, and then cancel their
+                # associated futures.
+                while True:
+                    try:
+                        work_item = self._work_queue.get_nowait()
+                    except queue.Empty:
+                        # Once the queue has been drained, send a wake-up to
+                        # prevent threads calling _work_queue.get(block=True)
+                        # from permanently blocking.
+                        self._work_queue.put(None)
```

Comment:
This seems unnecessary. Line 221 will wake up an idle thread, which will then add another `None` to the queue to cause any siblings to exit. If all the threads are currently doing work then they will check to see if the executor is shutdown after completing the work item and shutdown themselves. Am I missing a race?

Reply:
That was my impression initially as well, but it was causing a thread to indefinitely hang without this line (when `wait=True`). Attempting simply remove this line results in the following:

```py
[aeros:~/repos/aeros-cpython]$ ./python -m test test_concurrent_futures --match test_cancel_futures -v
== CPython 3.9.0a2+ (heads/bpo37224-fix-test__xxsubinterpreters.DestroyTests:6cbd9919ed, Jan 24 2020,) [GCC 9.2.0]
== Linux-5.4.8-arch1-1-x86_64-with-glibc2.30 little-endian
== cwd: /home/aeros/repos/aeros-cpython/build/test_python_157160
== CPU count: 4
== encodings: locale=UTF-8, FS=utf-8
0:00:00 load avg: 0.88 Run tests sequentially
0:00:00 load avg: 0.88 [1/1] test_concurrent_futures
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkProcessPoolShutdownTest) ... 0.32s ok
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkserverProcessPoolShutdownTest) ... 0.32s ok
test_cancel_futures (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolShutdownTest) ... 0.22s ok
test_cancel_futures (test.test_concurrent_futures.ThreadPoolShutdownTest) ... 
```
(before SIGINT, indefinite hang)

```py
angling was modified by test_concurrent_futures
  Before: {<weakref at 0x7f69554fedb0; to '_MainThread' at 0x7f6955f40a30>}
  After:  {<weakref at 0x7f69554fe270; to 'Thread' at 0x7f6955413430>, <weakref at 0x7f69554fecc0; to '_MainThread' at 0x7f6955f40a30>, <weakref at 0x7f69554fe2c0; to 'Thread' at 0x7f69553f1a30>, <weakref at 0x7f69554fe220; to 'Thread' at 0x7f6955509070>, <weakref at 0x7f69554fe1d0; to 'Thread' at 0x7f69553f15e0>, <weakref at 0x7f69554fe630; to 'Thread' at 0x7f69554131f0>} 


== Tests result: INTERRUPTED ==
Test suite interrupted by signal SIGINT.

1 test omitted:
    test_concurrent_futures

Total duration: 1 min 48 sec
Tests result: INTERRUPTED
```
(after SIGINT)

However, upon further experimentation, I discovered another way to fix this, by simply moving the `self._work_queue.put(None)` *after* draining the queue and cancelling the futures. I think the issue here was that the queue drain (in the `if cancel_futures:` block) was removing the `None` value from the queue before an idle `_worker` thread received it and caused the others to exit, leaving its `queue.get()` to block indefinitely. I'll adjust it accordingly.